src/gui/ui/scene_setup.cpp

```diff
@@ -197,9 +236,52 @@ setup(Scene& scene, shader::ShaderHandler& shader_handler, World& world) {
         chunks_shadow_program->data.push_back(chunk_mesh);
     }
 
+    voxel_utility::VoxelObject default_trees_voxel(
+        files::get_data_path() / "base" / "models" / "DefaultTree.qb"
+    );
+
+    auto mesh_trees = world::entity::ambient_occlusion_mesher(default_trees_voxel);
+
+    std::vector<glm::ivec4> model_matrices;
+    // generate positions of trees
+    for (size_t x = 0; x < world.get_terrain_main().get_X_MAX(); x += 40)
+        for (size_t y = 0; y < world.get_terrain_main().get_Y_MAX(); y += 40) {
+            size_t z = world.get_terrain_main().get_Z_solid(x, y) + 1;
+            if (z != 1) { // if the position of the ground is not zero
+                glm::ivec4 model(x, y, z, x / 40 + y / 40);
+                model_matrices.push_back(model);
+            }
+        }
+
+    // static because the mesh does not have moving parts
+    // this generates the buffer that holds the mesh data
+    auto gpu_trees_data =
+        std::make_shared<gpu_data::StaticMesh>(mesh_trees, model_matrices);
+    entity_shadow_program_execute->data.push_back(gpu_trees_data);
+    entity_render_program_execute->data.push_back(gpu_trees_data);
+
+    // attach the world objects to the render program
+    world::entity::ObjectHandler& object_handler =
+        world::entity::ObjectHandler::instance();
+    for (auto& [id, object] : object_handler.get_objects()) {
+        for (auto& mesh : object) {
+            // I'm so sorry for what I have done.
+            auto mesh_ptr = std::shared_ptr<world::entity::ModelController>(
```

Comment:
no no no no no

big UB and very big bug prone

Reply:
this is cpp comunity service. You create a shared ptr for the thing that takes a shared pointer, but pass the deconstructor you want to run. In this case that is nothing. The object is owned by a different thing, so it will be deconstructed when object handldr is removed.

Reply:
Stop being lazy and refactor your API.